qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,512 @@
+
+        Returns:
+            A callable that computes and returns the energy of the hamiltonian
+            of each parameter, and, optionally, the expectation
+
+        Raises:
+            RuntimeError: If the circuit is not parameterized (i.e. has 0 free parameters).
+            AlgorithmError: If operator was not provided.
+
+        """
+
+        num_parameters = self.ansatz.num_parameters
+        if num_parameters == 0:
+            raise RuntimeError("The ansatz must be parameterized, but has 0 free parameters.")
+
+        if operator is None:
+            raise AlgorithmError("The operator was never provided.")
+
+        if step > 1 and (len(prev_states) + 1) != step:
+            raise RuntimeError(
+                f"Passed previous states of the wrong size."
+                f"Passed array has length {str(len(prev_states))}"
+            )
+
+        self._check_operator_ansatz(operator)
+
+        prev_circs = []
+        for state in range(step - 1):
+            prev_circs.append(self.ansatz.bind_parameters(prev_states[state]))
+
+        def energy_evaluation(parameters):
+            estimator_job = self.estimator.run(
+                circuits=[self.ansatz], observables=[operator], parameter_values=[parameters]
+            )
+            estimator_result = estimator_job.result()
+            means = np.real(estimator_result.values)
+
+            if step > 1:
+                fidelity_job = self.fidelity.run(
+                    [self.ansatz] * len(prev_circs),
+                    prev_circs,
+                    [parameters] * len(prev_circs),
+                    [prev_states[:-1]],
+                )
+                costs = fidelity_job.result().fidelities
+
+                for (state, cost) in zip(range(step - 1), costs):
+                    means += np.real(self.betas[state] * cost)
+
+            if self.callback is not None:
+                variance = np.array([estimator_result.metadata[0].pop("variance", 0)])
+                shots = np.array([estimator_result.metadata[0].pop("shots", 0)])
+                estimator_error = np.sqrt(variance / shots)
+                for i, param_set in enumerate([parameters]):
+                    self._eval_count += 1
+                    self.callback(self._eval_count, param_set, means[i], estimator_error[i], step)
+            else:
+                self._eval_count += len(means)
+            return means if len(means) > 1 else means[0]
+
+        return energy_evaluation
+
+
+class VQDResult(VariationalResult, EigensolverResult):
+    """VQD Result."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._cost_function_evals = None
+
+    @property
+    def cost_function_evals(self) -> int | None:
+        """Returns number of cost optimizer evaluations"""
+        return self._cost_function_evals
+
+    @cost_function_evals.setter
+    def cost_function_evals(self, value: int) -> None:
```

Comment:
I don;t think the type is right here. Should be a list[int] since its the evals for each step (eigenvalue) as can be seen above in the compute method when the value that will end up here is built out (this was a single int for VQE but here should be a list - guess that was not spotted in the original code)

Reply:
Yes, I think it slipped under the radar in the original code. Fixing it